File: prototype/cogit/bisect.py
"""Bisect over thought history: find where reasoning first went wrong.

Contract: issues/COG-021.md (closes OQ-009). Probes are non-mutating —
the oracle inspects a candidate thought via env vars, no checkout happens.
"""

import os
import subprocess

from .errors import UserError

GOOD = "good"
BAD = "bad"
SKIP = "skip"


def command_runner(repo, command: str):
    """Oracle runner: exit 0 good, 125 skip, other < 128 bad, >= 128 aborts."""

    def run(thought_oid: str) -> str:
        thought = repo.store.read(thought_oid)
        env = {
            **os.environ,
            "COGIT_THOUGHT": thought_oid,
            "COGIT_MINDSET": thought["mindset"],
            "COGIT_REPO": os.path.dirname(repo.cogit_dir),
        }
        proc = subprocess.run(command, shell=True, env=env, capture_output=True)
        code = proc.returncode
        if code == 0:
            return GOOD
        if code == 125:
            return SKIP
        if 0 < code < 128:
            return BAD
        raise UserError(
            f"bisect-thought: oracle died (exit {code}) on {thought_oid}; aborting"
        )

    return run


def bisect_thought(repo, good: str, bad: str, runner):
    """Binary search for the first bad thought between good and bad.

    Assumes a monotonic predicate along the topological order of thoughts
    reachable from bad but not from good (MVP: linear histories).
    """
    good_oid = repo.resolve(good)
    bad_oid = repo.resolve(bad)
    ancestry_bad = repo._ancestry(bad_oid)
    if good_oid == bad_oid:
        raise UserError("bisect-thought: good and bad are the same thought")
    if good_oid not in ancestry_bad:
        raise UserError("bisect-thought: good must be an ancestor of bad")
    good_set = set(repo._ancestry(good_oid))
    candidates = {oid: t for oid, t in ancestry_bad.items() if oid not in good_set}
    order = repo._topo_oldest_first(candidates)
    # bad is the only candidate without descendants in the set -> emitted last
    assert order[-1] == bad_oid

    verdicts = {bad_oid: BAD}
    log = []
    lo, hi = 0, len(order) - 1
    while lo < hi:
        window = [i for i in range(lo, hi) if order[i] not in verdicts]
        if not window:
            return {
                "result": "inconclusive",
                "first_bad": None,
                "range": order[lo : hi + 1],
                "log": log,
                "candidates": len(order),
            }
        mid = window[len(window) // 2]
        verdict = runner(order[mid])
        log.append({"thought": order[mid], "verdict": verdict})
        verdicts[order[mid]] = verdict
        if verdict == SKIP:
            continue
        if verdict == BAD:
            hi = mid
        else:
            lo = mid + 1

    last_good = max(
        (i for i in range(hi) if verdicts.get(order[i]) == GOOD), default=-1
    )
    suspects = [
        order[i] for i in range(last_good + 1, hi) if verdicts.get(order[i]) == SKIP
    ]
    return {
        "result": "found",
        "first_bad": order[hi],
        "skipped_suspects": suspects,
        "log": log,
        "candidates": len(order),
    }

File: crates/cogit-core/src/bisect.rs
//! Bisect over thought history (COG-021, CQ-015). Probes are non-mutating:
//! the oracle receives COGIT_THOUGHT/COGIT_MINDSET/COGIT_REPO env vars.

use std::collections::{BTreeMap, BTreeSet};
use std::process::Command;

use serde_json::{json, Value};

use crate::error::{CoreError, Result};
use crate::repo::Repository;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Verdict {
    Good,
    Bad,
    Skip,
}

impl Verdict {
    fn as_str(self) -> &'static str {
        match self {
            Verdict::Good => "good",
            Verdict::Bad => "bad",
            Verdict::Skip => "skip",
        }
    }
}

/// Oracle runner: exit 0 good, 125 skip, other < 128 bad, >= 128 aborts.
pub fn run_command_oracle(repo: &Repository, command: &str, thought_oid: &str) -> Result<Verdict> {
    let thought = repo.store.read(thought_oid)?;
    let repo_root = repo.cogit_dir.parent().expect("repo root").to_path_buf();
    let status = Command::new("sh")
        .arg("-c")
        .arg(command)
        .env("COGIT_THOUGHT", thought_oid)
        .env("COGIT_MINDSET", thought["mindset"].as_str().unwrap_or(""))
        .env("COGIT_REPO", &repo_root)
        .status()?;
    match status.code() {
        Some(0) => Ok(Verdict::Good),
        Some(125) => Ok(Verdict::Skip),
        Some(code) if (1..128).contains(&code) => Ok(Verdict::Bad),
        other => Err(CoreError::User(format!(
            "bisect-thought: oracle died (exit {other:?}) on {thought_oid}; aborting"
        ))),
    }
}

/// Binary search for the first bad thought between good and bad.
pub fn bisect_thought<F>(repo: &Repository, good: &str, bad: &str, mut runner: F) -> Result<Value>
where
    F: FnMut(&str) -> Result<Verdict>,
{
    let good_oid = repo.resolve(good)?;
    let bad_oid = repo.resolve(bad)?;
    if good_oid == bad_oid {
        return Err(CoreError::User("bisect-thought: good and bad are the same thought".into()));
    }
    let ancestry_bad = repo.ancestry(&bad_oid)?;
    if !ancestry_bad.contains_key(&good_oid) {
        return Err(CoreError::User("bisect-thought: good must be an ancestor of bad".into()));
    }
    let good_set: BTreeSet<String> = repo.ancestry(&good_oid)?.into_keys().collect();
    let candidates: BTreeMap<String, Value> = ancestry_bad
        .into_iter()
        .filter(|(oid, _)| !good_set.contains(oid))
        .collect();
    let order = repo.topo_oldest_first(&candidates)?;
    assert_eq!(order.last(), Some(&bad_oid));

    let mut verdicts: BTreeMap<String, Verdict> = BTreeMap::from([(bad_oid.clone(), Verdict::Bad)]);
    let mut log: Vec<Value> = Vec::new();
    let (mut lo, mut hi) = (0usize, order.len() - 1);
    while lo < hi {
        let window: Vec<usize> = (lo..hi).filter(|i| !verdicts.contains_key(&order[*i])).collect();
        if window.is_empty() {
            return Ok(json!({
                "result": "inconclusive",
                "first_bad": null,
                "range": order[lo..=hi],
                "log": log,
                "candidates": order.len(),
            }));
        }
        let mid = window[window.len() / 2];
        let verdict = runner(&order[mid])?;
        log.push(json!({"thought": order[mid], "verdict": verdict.as_str()}));
        verdicts.insert(order[mid].clone(), verdict);
        match verdict {
            Verdict::Skip => continue,
            Verdict::Bad => hi = mid,
            Verdict::Good => lo = mid + 1,
        }
    }

    let last_good = (0..hi)
        .rev()
        .find(|i| verdicts.get(&order[*i]) == Some(&Verdict::Good))
        .map(|i| i as i64)
        .unwrap_or(-1);
    let suspects: Vec<&String> = ((last_good + 1) as usize..hi)
        .filter(|i| verdicts.get(&order[*i]) == Some(&Verdict::Skip))
        .map(|i| &order[i])
        .collect();
    Ok(json!({
        "result": "found",
        "first_bad": order[hi],
        "skipped_suspects": suspects,
        "log": log,
        "candidates": order.len(),
    }))
}
